Gmail app keeps relaunching the GPUProcess for each email you tap on

cdumez opened the pull request:
#### aab727be8e34adf97a438aa9f5a395d15f8e4c4e
<pre>
Gmail app keeps relaunching the GPUProcess for each email you tap on
<a href="https://bugs.webkit.org/show_bug.cgi?id=267664">https://bugs.webkit.org/show_bug.cgi?id=267664</a>
<a href="https://rdar.apple.com/121128159">rdar://121128159</a>

Reviewed by Per Arne Vollan and Brent Fulgham.

Gmail app keeps relaunching the GPUProcess for each email you tap on. This is
inefficient and results in a delay rendering the email.

The reason this happens is that the app keeps destroying a reconstructing
WebProcessPool objects. Given that the WebProcessPool object keeps alive the
GPUProcessProxy object, doing so would results in the GPUProcessProxy getting
reconstructed and thus the GPUProcess re-launching.

To address the issue, we now keep the GPUProcessProxy alive for a minute after
the last WebProcessPool object gets destroyed. This way, if a new WebProcessPool
gets re-constructed within a minute, it will just re-use the same GPU process.
Also, for apps that don&apos;t use WebKit much, the GPU process will still exit
eventually.

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::keepGPUProcessAliveTemporarily):
(WebKit::WebProcessPool::~WebProcessPool):
* Source/WebKit/UIProcess/WebProcessPool.h:

Canonical link: <a href="https://commits.webkit.org/273205@main">https://commits.webkit.org/273205@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/544e9919cd4c7f5d66ae98e90766510a3349ce41

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13494 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36678 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37355 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31308 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35812 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/15937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10588 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30267 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30886 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9978 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10053 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30963 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38627 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31476 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31275 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36108 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/10174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8065 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34070 "Found 1 new API test failure: /WebKitGTK/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11997 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/10722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4445 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11052 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -326,8 +326,31 @@ WebProcessPool::WebProcessPool(API::ProcessPoolConfiguration& configuration)
+#if ENABLE(GPU_PROCESS)
+    if (!m_gpuProcess)
+        return;
+
```
Should we `ASSERT` that we are on the main thread for this logic?

cdumez replied:
Sure, I'll add an assertion.

brentfulgham reviewed (approved):
Looks good. It's a shame the TestWebKitAPI setup has so much boilerplate. We should tidy that up someday.

pvollan commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -326,8 +326,31 @@ WebProcessPool::WebProcessPool(API::ProcessPoolConfiguration& configuration)
+    if (!m_gpuProcess)
+        return;
+
+    static constexpr Seconds durationToKeepGPUProcessAliveAfterDestruction { 1_min };
```
auto?

cdumez replied:
Ok.

pvollan reviewed (approved):
Nice! R=me.

cdumez committed:
Gmail app keeps relaunching the GPUProcess for each email you tap on
https://bugs.webkit.org/show_bug.cgi?id=267664
rdar://121128159

Reviewed by Per Arne Vollan and Brent Fulgham.

Gmail app keeps relaunching the GPUProcess for each email you tap on. This is
inefficient and results in a delay rendering the email.

The reason this happens is that the app keeps destroying a reconstructing
WebProcessPool objects. Given that the WebProcessPool object keeps alive the
GPUProcessProxy object, doing so would results in the GPUProcessProxy getting
reconstructed and thus the GPUProcess re-launching.

To address the issue, we now keep the GPUProcessProxy alive for a minute after
the last WebProcessPool object gets destroyed. This way, if a new WebProcessPool
gets re-constructed within a minute, it will just re-use the same GPU process.
Also, for apps that don't use WebKit much, the GPU process will still exit
eventually.

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::keepGPUProcessAliveTemporarily):
(WebKit::WebProcessPool::~WebProcessPool):
* Source/WebKit/UIProcess/WebProcessPool.h:

Canonical link: https://commits.webkit.org/273205@main
